src/Database/LSMTree/Internal/Run/Construction.hs

```diff
@@ -0,0 +1,416 @@
+        -- (7) the concatenation of all keys
+        <> mconcat [ BB.shortByteString (SBS ba#)
+                   | S.SerialisedKey (ByteArray ba#) <- toList pageKeys
+                   ]
+        -- (8) the concatenation of all values
+        <> mconcat [ BB.shortByteString (SBS ba#) -- TODO: wrong
+                   | RawValue _ _ (ByteArray ba#) <- toList pageValues
+                   ]
+  where
+    PageSize{..} = pageSize
+
+    n = pageSizeNumElems
+    b = pageSizeNumBlobs
+
+    -- Offset to component (5)
+    offKeyOffsets =
+         8                                         -- size of (1)
+      + (n + 63) `shiftR` 3 .&. complement 0x7     -- size of (2)
+      + (2 * n + 63) `shiftR` 3 .&. complement 0x7 -- size of (3)
+      + (4 + 8) * b                                -- size of (4)
+
+    -- Offset to component (7)
+    offKeys =
+        offKeyOffsets
+      + 2 * n                             -- size of (5)
+      + (if n == 1 then 6 else 2 * (n+1)) -- size of (6)
+
+    -- Thes values start right after the keys,
+    (pageKeyOffsets, offValues) =
+        case Seq.scanl (\o k -> o + sizeofKey16 k) offKeys pageKeys of
+          Empty      -> error "impossible"
+          kos :|> vo -> (kos, vo)
+
+    pageValueOffsets = case pageValues of
+      v :<| Empty -> Right (offValues, fromIntegral offValues + sizeofValue32 v)
+      _   -> Left (Seq.scanl (\o v -> o + sizeofValue16 v) offValues pageValues)
+
+{-------------------------------------------------------------------------------
+  Intermediate representation for a page
+-------------------------------------------------------------------------------}
+
+-- We might have preferred a representation for an incrementally constructed
+-- page that already builds up the raw bytes of a page. However, we do not know
+-- a priori what the offsets for most of the page components will be. These
+-- offsets depend on the number of keys\/values\/blob references, and the sizes
+-- of keys\/values, and we only find out these numbers as key-operation pairs
+-- are added incrementally. As such, we can only construct the raw page bytes
+-- once the current page has filled up.
+data PageIntermediate = PageIntermediate {
+    -- | (1) directory of components
+    pageSize           :: !PageSize
+    -- | (2) an array of 1-bit blob reference indicators
+  , pageBlobRefBitmap  :: !BitMap
+    -- | (3) an array of 2-bit operation types
+  , pageOperations     :: !CrumbMap
+    -- | (4) a pair of arrays of blob references
+  , pageBlobRefOffsets :: !(Seq Word64)
+    -- | (4) a pair of arrays of blob references, ctd
+  , pageBlobRefSizes   :: !(Seq Word32)
+    --   (5) key offsets will be computed when serialising the page
+    --   (6) value offsets will be computed when serialising the page
+    -- | (7) the concatenation of all keys
+  , pageKeys           :: !(Seq RawKey)
+    -- | (8) the concatenation of all values
+  , pageValues         :: !(Seq RawValue)
+  }
```

Comment:
I comment here on why I've used sequences, but one can also observe that even if we don't know the exact offsets, we do know bounds for how large each component can be. In particular, each component can not be larger than 4K. Of course, we can be more precise than that for each component, but even if we allocate a sufficiently sized `ByteArray` for each component, then we are pre-allocating quite a bit of memory for the worst case, while the average case might use less.

I haven't compute any numbers for the current approach and the approach I suggest in the paragraph above. Maybe I should, so we can compare them